# ExtraaLearn Project

## Context
The EdTech industry has been surging in the past decade immensely, and according to a forecast, the Online Education market would be worth $286.62bn by 2023 with a compound annual growth rate (CAGR) of 10.26% from 2018 to 2023. The modern era of online education has enforced a lot in its growth and expansion beyond any limit. Due to having many dominant features like ease of information sharing, personalized learning experience, transparency of assessment, etc, it is now preferable to traditional education.
In the present scenario due to the Covid-19, the online education sector has witnessed rapid growth and is attracting a lot of new customers. Due to this rapid growth, many new companies have emerged in this industry. With the availability and ease of use of digital marketing resources, companies can reach out to a wider audience with their offerings. The customers who show interest in these offerings are termed as leads. There are various sources of obtaining leads for Edtech companies, like

The customer interacts with the marketing front on social media or other online platforms. 
The customer browses the website/app and downloads the brochure
The customer connects through emails for more information.

The company then nurtures these leads and tries to convert them to paid customers. For this, the representative from the organization connects with the lead on call or through email to share further details.

## Objective
ExtraaLearn is an initial stage startup that offers programs on cutting-edge technologies to students and professionals to help them upskill/reskill. With a large number of leads being generated on a regular basis, one of the issues faced by ExtraaLearn is to identify which of the leads are more likely to convert so that they can allocate resources accordingly. You, as a data scientist at ExtraaLearn, have been provided the leads data to:

Analyze and build an ML model to help identify which leads are more likely to convert to paid customers, 
Find the factors driving the lead conversion process
Create a profile of the leads which are likely to convert


## Data Description
The data contains the different attributes of leads and their interaction details with ExtraaLearn. The detailed data dictionary is given below.
Data Dictionary

ID: ID of the lead
age: Age of the lead
current_occupation: Current occupation of the lead. Values include 'Professional','Unemployed',and 'Student'
first_interaction: How did the lead first interacted with ExtraaLearn. Values include 'Website', 'Mobile App'
profile_completed: What percentage of profile has been filled by the lead on the website/mobile app. Values include Low - (0-50%), Medium - (50-75%), High (75-100%)
website_visits: How many times has a lead visited the website
time_spent_on_website: Total time spent on the website
page_views_per_visit: Average number of pages on the website viewed during the visits.
last_activity: Last interaction between the lead and ExtraaLearn.

Email Activity: Seeking for details about program through email, Representative shared information with lead like brochure of program , etc 
Phone Activity: Had a Phone Conversation with representative, Had conversation over SMS with representative, etc
Website Activity: Interacted on live chat with representative, Updated profile on website, etc


print_media_type1: Flag indicating whether the lead had seen the ad of ExtraaLearn in the Newspaper.

print_media_type2: Flag indicating whether the lead had seen the ad of ExtraaLearn in the Magazine.
digital_media: Flag indicating whether the lead had seen the ad of ExtraaLearn on the digital platforms.
educational_channels: Flag indicating whether the lead had heard about ExtraaLearn in the education channels like online forums, discussion threads, educational websites, etc.
referral: Flag indicating whether the lead had heard about ExtraaLearn through reference.
status: Flag indicating whether the lead was converted to a paid customer or not.

## Importing necessary libraries and data

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#preprocessing
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import RobustScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

#Algorithms
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

#Metrics
from sklearn import metrics
from sklearn.metrics import confusion_matrix, classification_report, precision_recall_curve, recall_score
from sklearn.metrics import precision_score, make_scorer


from sklearn import tree

#To ignore warnings
import warnings
warnings.filterwarnings("ignore")

## Data Overview

Observations
Sanity checks

In [ ]:
#Loading Dataset
df = pd.read_csv(r'C:\Users\Administrator\Desktop\python projects\ExtraaLearn.csv')
df.head()

ID
age
current_occupation
first_interaction
profile_completed
website_visits
time_spent_on_website
page_views_per_visit
last_activity
print_media_type1
print_media_type2
digital_media
educational_channels
referral
status




0
EXT001
57
Unemployed
Website
High
7
1639
1.861
Website Activity
Yes
No
Yes
No
No
1


1
EXT002
56
Professional
Mobile App
Medium
2
83
0.320
Website Activity
No
No
No
Yes
No
0


2
EXT003
52
Professional
Website
Medium
3
330
0.074
Website Activity
No
No
Yes
No
No
0


3
EXT004
53
Unemployed
Website
High
4
464
2.057
Website Activity
No
No
No
No
No
1


4
EXT005
23
Student
Website
High
4
600
16.914
Email Activity
No
No
No
No
No
0


In [ ]:
df.shape

(4612, 15)


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4612 entries, 0 to 4611
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID                     4612 non-null   object 
 1   age                    4612 non-null   int64  
 2   current_occupation     4612 non-null   object 
 3   first_interaction      4612 non-null   object 
 4   profile_completed      4612 non-null   object 
 5   website_visits         4612 non-null   int64  
 6   time_spent_on_website  4612 non-null   int64  
 7   page_views_per_visit   4612 non-null   float64
 8   last_activity          4612 non-null   object 
 9   print_media_type1      4612 non-null   object 
 10  print_media_type2      4612 non-null   object 
 11  digital_media          4612 non-null   object 
 12  educational_channels   4612 non-null   object 
 13  referral               4612 non-null   object 
 14  status                 4612 non-null   int64  
dtypes: f

In [ ]:
df.isnull().sum()

ID                       0
age                      0
current_occupation       0
first_interaction        0
profile_completed        0
website_visits           0
time_spent_on_website    0
page_views_per_visit     0
last_activity            0
print_media_type1        0
print_media_type2        0
digital_media            0
educational_channels     0
referral                 0
status                   0
dtype: int64


#### Observations:
The dataset has 4612 rows and 15 columns.
It has no null values.
datatypes: float64(1), int64(4), object(10)

In [ ]:
df.nunique()

ID                       4612
age                        46
current_occupation          3
first_interaction           2
profile_completed           3
website_visits             27
time_spent_on_website    1623
page_views_per_visit     2414
last_activity               3
print_media_type1           2
print_media_type2           2
digital_media               2
educational_channels        2
referral                    2
status                      2
dtype: int64


### Observation:

The column ID is unique for each lead. so we can drop the column as it wouls not add any value to our analysis.

In [ ]:
df = df.drop(['ID'],axis=1)

## Exploratory Data Analysis (EDA)

EDA is an important part of any project involving data.
It is important to investigate and understand the data better before building a model with it.
A few questions have been mentioned below which will help you approach the analysis in the right manner and generate insights from the data.
A thorough analysis of the data, in addition to the questions mentioned below, should be done.

Questions

Leads will have different expectations from the outcome of the course and the current occupation may play a key role in getting them to participate in the program. Find out how current occupation affects lead status.
The company's first impression on the customer must have an impact. Do the first channels of interaction have an impact on the lead status? 
The company uses multiple modes to interact with prospects. Which way of interaction works best? 
The company gets leads from various channels such as print media, digital media, referrals, etc. Which of these channels have the highest lead conversion rate?
People browsing the website or mobile application are generally required to create a profile by sharing their personal data before they can access additional information.Does having more details about a prospect increase the chances of conversion?

## Data Preprocessing

Missing value treatment (if needed)
Feature engineering (if needed)
Outlier detection and treatment (if needed)
Preparing data for modeling 
Any other preprocessing steps (if needed)

In [ ]:
df.head()

age
current_occupation
first_interaction
profile_completed
website_visits
time_spent_on_website
page_views_per_visit
last_activity
print_media_type1
print_media_type2
digital_media
educational_channels
referral
status




0
57
Unemployed
Website
High
7
1639
1.861
Website Activity
Yes
No
Yes
No
No
1


1
56
Professional
Mobile App
Medium
2
83
0.320
Website Activity
No
No
No
Yes
No
0


2
52
Professional
Website
Medium
3
330
0.074
Website Activity
No
No
Yes
No
No
0


3
53
Unemployed
Website
High
4
464
2.057
Website Activity
No
No
No
No
No
1


4
23
Student
Website
High
4
600
16.914
Email Activity
No
No
No
No
No
0


In [ ]:
#Numerical columns:
num_cols = ['age','website_visits','time_spent_on_website','page_views_per_visit'] 
#Categorical columns:
cat_cols = ['current_occupation','first_interaction','profile_completed','last_activity','print_media_type1','print_media_type2','digital_media','educational_channels','referral']

# univariate analysis:

#### Numerival Variables:

In [ ]:
#statiscal summary:
df[num_cols].describe().T

count
mean
std
min
25%
50%
75%
max




age
4612.0
46.201214
13.161454
18.0
36.00000
51.000
57.00000
63.000


website_visits
4612.0
3.566782
2.829134
0.0
2.00000
3.000
5.00000
30.000


time_spent_on_website
4612.0
724.011275
743.828683
0.0
148.75000
376.000
1336.75000
2537.000


page_views_per_visit
4612.0
3.026126
1.968125
0.0
2.07775
2.792
3.75625
18.434


In [ ]:
#Histograms:
df[num_cols].hist(figsize=(8,8))
plt.show()

In [ ]:
for col in num_cols:
    plt.figure(figsize=(8, 4))
    sns.boxplot(data=df, y=col)
    plt.title(f"Boxplot of {col}")
    plt.show()

### Observations:
Average age of leads is around 47 years it has a wide range, starting 18 to 63 which indicates good age diversity of leads.
website_visits is skewed to right average website_visits of leads is around 3 times.
page_views_per_visit is also skewed to right average page_views_per_visit is around 3 pages.

## Univariate analysis:

#### categorical variables:

In [ ]:
print(df['status'].value_counts(normalize=True))

0    0.701431
1    0.298569
Name: status, dtype: float64


In [ ]:
for i in cat_cols:
    print(df[i].value_counts(normalize=True))
    print('_'*50)

Professional    0.567216
Unemployed      0.312446
Student         0.120338
Name: current_occupation, dtype: float64
__________________________________________________
Website       0.551171
Mobile App    0.448829
Name: first_interaction, dtype: float64
__________________________________________________
High      0.490893
Medium    0.485906
Low       0.023200
Name: profile_completed, dtype: float64
__________________________________________________
Email Activity      0.493929
Phone Activity      0.267563
Website Activity    0.238508
Name: last_activity, dtype: float64
__________________________________________________
No     0.892238
Yes    0.107762
Name: print_media_type1, dtype: float64
__________________________________________________
No     0.94948
Yes    0.05052
Name: print_media_type2, dtype: float64
__________________________________________________
No     0.885733
Yes    0.114267
Name: digital_media, dtype: float64
__________________________________________________
No     0.84

In [ ]:
status_count = df['status'].value_counts(normalize=True)
print(status_count)

0    0.701431
1    0.298569
Name: status, dtype: float64


### Observations:
The lead to paid customer rate is around 30%.
Approximately 56% of the leads falls under the category of professional(currently working) in current_occupation column. which indicates more than half of leads are currently employed.
31% of unemployed and 12% of students.
The first_interaction of leads are 55% on website and 44% on mobile app.
50% of the leads activity is Email_Activity. 
of all the media types educational_channels and digital_media both influences people comparitively higher but its still only 12 to 15%
The leads came from referral seems to have conversion rate even if the overall percentage is low.

#### Bivariate and Multivariate analysis:

In [ ]:
for column in df.columns:
    if column != "status":
        plt.figure(figsize=(6,3))
        if df[column].dtype == "object":
            sns.countplot(data=df, x=column, hue="status")
        else: 
            sns.boxplot(data=df,x="status", y=column)
        plt.title(f"Comparison of 'status' with '{column}'")
        plt.show()

In [ ]:
status_percentage = df[df["status"] == 1].groupby("current_occupation")["status"].count() / df.groupby("current_occupation")["status"].count() * 100
print(status_percentage)
plt.figure(figsize=(8, 4))
ax = sns.barplot(data=status_percentage.reset_index(), x="current_occupation", y="status", palette="Set2")
plt.xlabel("current_occupation")
plt.ylabel("Percentage of leads with status = 1")
plt.title("Comparison of status with current_occupation")

for p in ax.patches:
    ax.annotate(f'{p.get_height():.2f}%', (p.get_x() + p.get_width() / 2, p.get_height()), ha='center', va='center')

plt.show()

current_occupation
Professional    35.512232
Student         11.711712
Unemployed      26.578765
Name: status, dtype: float64


In [ ]:
for column in df.columns:
    if column != "status":
        percentage = df[df["status"] == 1][column].value_counts(normalize=True) * 100
        print(f"Percentage of leads with status = 1 for {column}:")
        print(percentage)
        print()

Percentage of leads with status = 1 for age:
57    9.876543
59    8.787219
56    8.714597
58    8.714597
60    5.519245
55    5.228758
32    3.994190
53    2.904866
50    2.687001
45    2.396514
34    2.251271
46    2.106028
41    1.888163
44    1.815541
47    1.815541
49    1.815541
43    1.742919
42    1.742919
48    1.597676
52    1.597676
54    1.525054
37    1.525054
36    1.525054
35    1.452433
39    1.379811
51    1.379811
40    1.161946
30    1.016703
31    0.944081
38    0.944081
20    0.871460
29    0.871460
18    0.798838
62    0.798838
63    0.798838
24    0.798838
33    0.726216
28    0.653595
22    0.653595
61    0.653595
21    0.653595
19    0.508351
23    0.435730
26    0.290487
27    0.290487
25    0.145243
Name: age, dtype: float64

Percentage of leads with status = 1 for current_occupation:
Professional    67.465505
Unemployed      27.814089
Student          4.720407
Name: current_occupation, dtype: float64

Percentage of leads with status = 1 for first_interaction:

In [ ]:
for i in cat_cols:
    for j in num_cols:
        sns.barplot(data=df, x=i, y=j)
        plt.show()

In [ ]:
#Multivariate analysis:
correlation_matrix = df.corr()
sns.heatmap(correlation_matrix, annot=True)
plt.show()

#### Observations:
The leads with age above 44 are comparitively likely to converted to paid customer. The average age of converted paid customer is around 54.
1.About 67% of paid customers are currently working(profession) leads.
comparitively the leads converted to paid customers from the category of profession(currently working) in current_occupation are higher in number than others categories(student and unemployed).
2.About 45% of paid customers have website as their first interaction.
5.About 68% of paid customers profile_completed status is high.which indicates having more details about a prospect increase the chances of conversion. and leads using websites tend to have more 'high' status of profile completion.
The column website_visits has more number of outliers
the leads converted to paid customers seems like spend more time on website (time_spent_on_website)
The column page_views_per_visit also has more number of outliers.
3.About 50% of the paid customers has their last_activity as Email Activity.
4.comparitively the educational channel and digital media has a bit more influence on leads to become paid customers although its only 12 to 15%.
6.The leads came from referral seems to have conversion rate even if the overall percentage is low.

## EDA

It is a good idea to explore the data once again after manipulating it.

### Data Preprocessing

#### Observations:
1.There is no null values in data set. so there is no need to be treated.
2.#### Feature Engineering:
   The column ID has unique values which would not help the analysis so its been dropped.
3.The columns website_visits and  page_views_per_visit has more outliers.

### Seperating target and other variables:

In [ ]:
# Separating target variable and other variables

Y= df.status
X= df.drop(columns = ['status'])

### Splitting the dataset into train and test data:

In [ ]:
# Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.3, random_state = 1, stratify = Y)

### Scaling the data:

In [ ]:
#scaling
# seperating numerical and categorical columns
numerical_cols = X_train.select_dtypes(include='number').columns
categorical_cols = X_train.select_dtypes(include='object').columns

# preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', RobustScaler(), numerical_cols),
        ('cat', OneHotEncoder(), categorical_cols)
    ])
pipeline = Pipeline(steps=[('preprocessor', preprocessor)])

In [ ]:
# Apply preprocessing to training and testing data
X_train_processed = pipeline.fit_transform(X_train)
X_test_processed = pipeline.transform(X_test)

# Concatenate the processed numerical and categorical features
X_train_final = pd.DataFrame(X_train_processed, columns=list(numerical_cols) + list(preprocessor.named_transformers_['cat'].get_feature_names_out(categorical_cols)))
X_test_final = pd.DataFrame(X_test_processed, columns=list(numerical_cols) + list(preprocessor.named_transformers_['cat'].get_feature_names_out(categorical_cols)))

# Check the final processed data
print(X_train_final.head())
print(X_test_final.head())

age  website_visits  time_spent_on_website  page_views_per_visit  \
0  0.285714       -0.666667               0.168443             -0.396616   
1 -0.761905       -0.333333               1.279098              1.333531   
2  0.095238        0.666667               0.965164             -0.440552   
3  0.047619        1.333333              -0.296311              0.900698   
4  0.285714       -1.000000              -0.308607             -1.701054   

   num_details  current_occupation_Professional  current_occupation_Student  \
0          0.0                              1.0                         0.0   
1          0.0                              0.0                         0.0   
2          0.0                              0.0                         0.0   
3          0.0                              1.0                         0.0   
4          0.0                              1.0                         0.0   

   current_occupation_Unemployed  first_interaction_Mobile App  \
0         

In [ ]:
# Creating metric function 
def metrics_score(actual, predicted):
    print(classification_report(actual, predicted))

    cm = confusion_matrix(actual, predicted)
    plt.figure(figsize=(8,5))
    
    sns.heatmap(cm, annot=True,  fmt='.2f', xticklabels=['Not Converted', 'Converted'], yticklabels=['Not Converted', 'Converted'])
    plt.ylabel('Actual')
    plt.xlabel('Predicted')
    plt.show()

## Building a Decision Tree model

In [ ]:
# Building decision tree model
class_weights = {0: 1/0.701431, 1: 1/0.298569}
dt = DecisionTreeClassifier(class_weight=class_weights, random_state=1)

In [ ]:
# Fitting decision tree model
dt.fit(X_train_final, y_train)

DecisionTreeClassifier(class_weight={0: 1.4256569783770605,
                                     1: 3.3493095398383628},
                       random_state=1)In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.DecisionTreeClassifierDecisionTreeClassifier(class_weight={0: 1.4256569783770605,
                                     1: 3.3493095398383628},
                       random_state=1)


In [ ]:
# Checking performance on the training dataset
y_train_pred_dt = dt.predict(X_train_final)

metrics_score(y_train, y_train_pred_dt)

precision    recall  f1-score   support

           0       1.00      1.00      1.00      2264
           1       1.00      1.00      1.00       964

    accuracy                           1.00      3228
   macro avg       1.00      1.00      1.00      3228
weighted avg       1.00      1.00      1.00      3228


### Observation:
The Decision tree is giving a 100% score for all metrics on the training dataset.

In [ ]:
# Checking performance on the test dataset
y_test_pred_dt = dt.predict(X_test_final)

metrics_score(y_test, y_test_pred_dt)

precision    recall  f1-score   support

           0       0.85      0.87      0.86       971
           1       0.67      0.64      0.65       413

    accuracy                           0.80      1384
   macro avg       0.76      0.75      0.76      1384
weighted avg       0.80      0.80      0.80      1384


Observations:
The Decision Tree works well on the training data but not so well on the test data as the precision is 0.67 in comparison to 1 for the training dataset, i.e., the Decision Tree is overfitting the training data.
This indicates that the model's performance in correctly identifying converted leads is relatively lower compared to class 0.
Based on these observations, the model appears to perform well in identifying leads that were not converted to paid customers (class 0) but may have room for improvement in correctly identifying converted leads (class 1).

In [ ]:
# Plot the feature importance

importances = dt.feature_importances_
feature_names = X_train_final.columns
indices = importances.argsort()[::-1]

plt.figure(figsize=(10, 6))
plt.bar(range(len(importances)), importances[indices])
plt.xticks(range(len(importances)), feature_names[indices], rotation=90)
plt.xlabel("Feature")
plt.ylabel("Importance")
plt.title("Feature Importances")
plt.show()

### Observations:
According to the Decision Tree, time_spent_on_website followed by first_interaction_Website and page_views_per_visit  are the most important features

## Do we need to prune the tree?

since the Deciion tree is overfitting the training data pruning has to be done

In [ ]:
class_weights = {0: 1/0.701431, 1: 1/0.298569}
dt = DecisionTreeClassifier(class_weight=class_weights, random_state=1)

#parameter grid for hyperparameter tuning
param_grid = {'ccp_alpha': [0.001, 0.002, 0.003, 0.004, 0.005]}  # Adjust the range of ccp_alpha as needed

# Perform GridSearchCV to find the optimal ccp_alpha
grid_search = GridSearchCV(dt, param_grid, cv=5)
grid_search.fit(X_train_final, y_train)

# Get the best ccp_alpha value
best_ccp_alpha = grid_search.best_params_['ccp_alpha']

# Fit the decision tree with the best ccp_alpha value
pruned_dt = DecisionTreeClassifier(ccp_alpha=best_ccp_alpha, random_state=1)
pruned_dt.fit(X_train_final, y_train)

DecisionTreeClassifier(ccp_alpha=0.002, random_state=1)In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.DecisionTreeClassifierDecisionTreeClassifier(ccp_alpha=0.002, random_state=1)


In [ ]:
y_pred = pruned_dt.predict(X_test_final)
metrics_score(y_test, y_pred)

precision    recall  f1-score   support

           0       0.88      0.93      0.91       971
           1       0.82      0.71      0.76       413

    accuracy                           0.87      1384
   macro avg       0.85      0.82      0.84      1384
weighted avg       0.87      0.87      0.87      1384


#### Observations:
The overall accuracy of the model on the test data is 0.87, compariritively higher. indicating that it correctly classified 87% of the leads.
The model achieved a precision of 0.82, recall of 0.71, and F1-score of 0.76. This suggests that the model's performance in correctly identifying converted leads is relatively lower compared to class 0, with a slightly lower precision and recall.

In [ ]:
# lets plot the feature importances:
importances = pruned_dt.feature_importances_
feature_names = X_train_final.columns
indices = importances.argsort()[::-1]

plt.figure(figsize=(10, 6))
plt.bar(range(len(importances)), importances[indices])
plt.xticks(range(len(importances)), feature_names[indices], rotation=90)
plt.xlabel("Feature")
plt.ylabel("Importance")
plt.title("Feature Importances")
plt.show()

#### Observations:
The feature importances has changed here, the current_occupation_profssional gained some importance in the pruned model compared to the old decision tree.

## Building a Random Forest model

In [ ]:
class_weights = {0: 1/0.701431, 1: 1/0.298569}
rf = RandomForestClassifier(n_estimators=100, class_weight=class_weights, random_state=1)

# Fit the Random Forest model on the training data
rf.fit(X_train_final, y_train)

# Make predictions on the test data
y_pred = rf.predict(X_test_final)

In [ ]:
# Checking performance on the training data
y_pred_train_rf = rf.predict(X_train_final)
metrics_score(y_train, y_pred_train_rf)

precision    recall  f1-score   support

           0       1.00      1.00      1.00      2264
           1       1.00      1.00      1.00       964

    accuracy                           1.00      3228
   macro avg       1.00      1.00      1.00      3228
weighted avg       1.00      1.00      1.00      3228


Observation:
For all the metrics in the training dataset, the Random Forest gives a 100% score.

In [ ]:
# Checking performance on the testing data
y_pred_test_rf = rf.predict(X_test_final)
metrics_score(y_test, y_pred_test_rf)

precision    recall  f1-score   support

           0       0.89      0.92      0.90       971
           1       0.80      0.72      0.76       413

    accuracy                           0.86      1384
   macro avg       0.84      0.82      0.83      1384
weighted avg       0.86      0.86      0.86      1384


#### Observations:
For class 0 (leads not converted to paid customers), the model achieved a precision of 0.89, recall of 0.92, and F1-score of 0.90. This indicates that the model performed well in correctly identifying leads that were not converted to paid customers, with a high precision and recall.
For class 1 (leads converted to paid customers), the model achieved a precision of 0.80, recall of 0.72, and F1-score of 0.76. This suggests that the model's performance in correctly identifying converted leads is relatively lower compared to class 0, with a lower precision and recall.
The overall accuracy of the model on the test data is 0.86, indicating that it correctly classified 86% of the leads.

In [ ]:
importances = rf.feature_importances_
feature_names = X_train_final.columns
indices = importances.argsort()[::-1]

plt.figure(figsize=(10, 6))
plt.bar(range(len(importances)), importances[indices])
plt.xticks(range(len(importances)), feature_names[indices], rotation=90)
plt.xlabel("Feature")
plt.ylabel("Importance")
plt.title("Feature Importances")
plt.show()

## Do we need to prune the tree?

In the case of a Random Forest model, which is an ensemble of decision trees, pruning is not typically necessary. Random Forest inherently addresses overfitting by combining multiple trees and aggregating their predictions. The ensemble nature of Random Forest helps to reduce the individual tree's overfitting tendencies and create a more robust and generalized model.

### Observations:
we have tried decision trees and random forest models, and were able to identify  the key values involved with high conversion ratte of leads.
Random Forest Model: The Random Forest model and pruned decission tree: achieved high accuracy, precision, and recall on the training data. However, it showed some signs of overfitting, as indicated by the drop in performance on the test data. Consider further model refinement, such as adjusting hyperparameters, collecting additional relevant features, or exploring ensemble techniques like bagging or boosting, to improve the model's generalization performance.
but we can improve the model performance by tuning the hyper parameters.

### Tuning hyper parameters

In [ ]:
# Choose the type of classifier
# Grid of parameters to choose from
param_grid = {
    'n_estimators': [100, 200, 300],  # Number of trees in the forest
    'max_depth': [None, 5, 10],       # Maximum depth of the trees
    'min_samples_split': [2, 5, 10],  # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 2, 4],    # Minimum number of samples required to be at a leaf node
    'max_features': ['sqrt', 'log2']  # Number of features to consider when looking for the best split
}

grid_search = GridSearchCV(estimator=rf_tuned, param_grid=param_grid, scoring=make_scorer(precision_score), cv=5)

grid_search.fit(X_train_final, y_train)

# Get the best parameters and best score from the grid search
best_params = grid_search.best_params_
best_score = grid_search.best_score_

print("Best Parameters:", best_params)
print("Best Precision Score:", best_score)

Best Parameters: {'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Best Precision Score: 0.7985382159347411


In [ ]:
# Set the classifier to the best combination of parameters
#rf_tuned = grid_search.best_estimator_

In [ ]:
class_weights = {0: 1/0.701431, 1: 1/0.298569}
rf_tuned = RandomForestClassifier(class_weight = class_weights, n_estimators=100, max_depth=None, min_samples_split=2, min_samples_leaf=1, max_features='sqrt', random_state=1)
rf_tuned.fit(X_train_final, y_train)

RandomForestClassifier(class_weight={0: 1.4256569783770605,
                                     1: 3.3493095398383628},
                       random_state=1)In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.RandomForestClassifierRandomForestClassifier(class_weight={0: 1.4256569783770605,
                                     1: 3.3493095398383628},
                       random_state=1)


In [ ]:
# Checking performance on the training data
y_pred_train_rf_tuned = rf_tuned.predict(X_train_final)

metrics_score(y_train, y_pred_train_rf_tuned)

precision    recall  f1-score   support

           0       1.00      1.00      1.00      2264
           1       1.00      1.00      1.00       964

    accuracy                           1.00      3228
   macro avg       1.00      1.00      1.00      3228
weighted avg       1.00      1.00      1.00      3228


In [ ]:
# Checking performance on the test data
y_pred_test_rf_tuned = rf_tuned.predict(X_test_final)

metrics_score(y_test, y_pred_test_rf_tuned)

precision    recall  f1-score   support

           0       0.89      0.92      0.90       971
           1       0.80      0.72      0.76       413

    accuracy                           0.86      1384
   macro avg       0.84      0.82      0.83      1384
weighted avg       0.86      0.86      0.86      1384


### Observations:
Based on the evaluation metrics you provided it appears that the Random Forest classifier with the best parameters achieved relatively good performance on the test data. Here's a breakdown of the metrics:
Precision: The model achieved a precision of 0.89 for class 0 and 0.80 for class 1. Precision measures the proportion of correctly predicted positive instances out of the total predicted positive instances. A higher precision indicates that the model has a low rate of false positives.
Recall: The model achieved a recall of 0.92 for class 0 and 0.72 for class 1. Recall measures the proportion of correctly predicted positive instances out of the total actual positive instances. A higher recall indicates that the model has a low rate of false negatives.
F1-score: The model achieved an F1-score of 0.90 for class 0 and 0.76 for class 1. The F1-score is the harmonic mean of precision and recall, providing a balanced measure of model performance.
Accuracy: The model achieved an overall accuracy of 0.86, which measures the proportion of correctly predicted instances out of the total instances. It indicates the overall effectiveness of the model.

## Actionable Insights and Recommendations

Based on the analysis and model evaluation, here are some actionable insights and recommendations:
1.Profile Completion: The analysis showed that leads with a higher profile completion rate tend to have a higher likelihood of conversion. Encourage leads to provide more detailed information and incentives to complete their profiles, as it can increase the chances of conversion.
2.Website Visits and Engagement: Leads who have higher website visits and spend more time on the website tend to show a higher conversion rate. Focus on improving website user experience, providing valuable content, and optimizing lead engagement to increase the chances of conversion.
3.Page Views per Visit: The analysis revealed that leads with a higher number of page views per visit are more likely to convert. This indicates that leads who explore more pages on the website are more engaged and interested. Consider optimizing website navigation and providing relevant content to encourage leads to explore multiple pages, thereby increasing conversion opportunities.
4.Current Occupation: The analysis revealed that the current occupation of leads has an impact on conversion. Professional leads have a higher conversion rate compared to students and unemployed leads. Tailor marketing strategies and messaging to target professionals effectively and address their specific needs and pain points.
5.Referral Program: The analysis showed that leads who came through referrals have a higher conversion rate. Leverage the power of referrals by implementing a referral program that incentivizes existing customers to refer new leads. This can help increase the number of high-quality leads and improve conversion rates.
6.Random Forest Model: The Random Forest model achieved high accuracy, precision, and recall on the training data. However, it showed some signs of overfitting, as indicated by the drop in performance on the test data. Consider further model refinement, such as adjusting hyperparameters, collecting additional relevant features, or exploring ensemble techniques like bagging or boosting, to improve the model's generalization performance.